### DISCUSSION TITLE
MeshInstancing (How the drawcalls should behave `DrawMesh, DrawMeshInstanced`)

### DISCUSSION THREAD

**Discussion by MrDiver** [raysan5/raylib#1950] [Q&A] [answered]
Hello Everyone,

in the last couple of days, I started using Raylib very heavily because it was a beautiful and quick alternative for rendering with OpenGL without having to write a whole library yourself. After going through the code for a while and finding ways to implement some basic shading techniques. When I tried to do instanced rendering, I quickly ran in to some problems.

When using the `DrawMeshInstanced(...)` Function with only one instance, it doesn't use the instance rendering mode. This makes it very weird when using the same shader on draw calls for one and more instances because it changes from using the attribute to using the uniform.

I ran in to this while trying to implement deferred rendering.

Now, the important question is, why is the non instancing part not actually in the `DrawMesh(...)` Function?
So that the `DrawMeshInstanced(...)` Function is always using instanced rendering, no matter how many instances.
In my opinion, it seems perfectly reasonable to have instanced rendering and normal rendering separated in two different functions so that the user can decide on which to use and doesn't get surprised when it suddenly changes to the other method of rendering.

If this was discussed in an earlier issue, then I am sorry that I didn't find it.

If not, I would suggest discussing this point here to maybe help improve how this process works.

Referenced Code:
https://github.com/raysan5/raylib/blob/6575d31379045f34fa7d23b79d9adea2f18899e5/src/models.c#L974-L977
Inside `DrawMeshInstanced(...)`
https://github.com/raysan5/raylib/blob/6575d31379045f34fa7d23b79d9adea2f18899e5/src/models.c#L1018
https://github.com/raysan5/raylib/blob/6575d31379045f34fa7d23b79d9adea2f18899e5/src/models.c#L1067

PS: _Yes, I am aware that there's also the possibility to write your own Instanced Mesh rendering function that wouldn't do this, but this is not what I want to discuss with this._

**Comment by 64bit-Paperclip** [comment]
I think the idea is spot on for how things should be handled.

**Comment by raysan5** [comment, marked as answer]
@MrDiver This issue has been solved in https://github.com/raysan5/raylib/issues/1958
